src/network_analysis.jl

```diff
@@ -938,3 +942,79 @@ end
+"""
+
+function satisfiesdeficiencyone(rn::ReactionSystem)
+    complexes, D = reactioncomplexes(rn)
+    δ = deficiency(rn)
+    δ_l = linkagedeficiencies(rn)
+
+    lcs = linkageclasses(rn)
+    tslcs = terminallinkageclasses(rn)
+
+    # Check the conditions for the deficiency one theorem: 
+    #   1) the deficiency of each individual linkage class is at most 1; 
+    #   2) the sum of the linkage deficiencies is the total deficiency, and 
+    #   3) there is only one terminal linkage class per linkage class. 
+    all(<=(1), δ_l) && (sum(δ_l) == δ) && (length(lcs) == length(tslcs))
+end
+
+"""
+    satisfiesdeficiencyzero(rn::ReactionSystem)
+
+    Check if a reaction network obeys the conditions of the deficiency zero theorem, which ensures that there is only one equilibrium for every positive stoichiometric compatibility class, this equilibrium is asymptotically stable, and this equilibrium is complex balanced.
+"""
+
+function satisfiesdeficiencyzero(rn::ReactionSystem)
+    δ = deficiency(rn)
+    δ == 0 && isweaklyreversible(rn, subnetworks(rn))
+end
+
+"""
+    robustspecies(rn::ReactionSystem)
+
+    Return a vector of indices corresponding to species that are concentration robust, i.e. for every positive equilbrium, the concentration of species s will be the same. 
+
+    Note: This function currently only works for networks of deficiency one, and is not currently guaranteed to return *all* the concentration-robust species in the network. Any species returned by the function will be robust, but this may not include all of them. Use with caution. Support for higher deficiency networks and necessary conditions for robustness will be coming in the future.  
+"""
+
+function robustspecies(rn::ReactionSystem)
+    complexes, D = reactioncomplexes(rn)
+    nps = get_networkproperties(rn)
+
+    if deficiency(rn) != 1
+        error("This algorithm currently only checks for robust species in networks with deficiency one.")
+    end
+
+    # A species is concnetration robust in a deficiency one network if there are two non-terminal complexes (i.e. complexes 
+    # belonging to a linkage class that is not terminal) that differ only in species s (i.e. their difference is some 
+    # multiple of s. (A + B, A) differ only in B. (A + 2B, B) differ in both A and B, since A + 2B - B = A + B). 
+
+    if !nps.checkedrobust
+        tslcs = terminallinkageclasses(rn)
+        Z = complexstoichmat(rn)
+
+        # Find the complexes that do not belong to a terminal linkage class 
+        nonterminal_complexes = deleteat!([1:length(complexes);], vcat(tslcs...))
+        robust_species = Int64[]
+
+        for (c_s, c_p) in collect(Combinatorics.combinations(nonterminal_complexes, 2))
+            # Check the difference of all the combinations of complexes. The support is the set of indices that are non-zero 
+            supp = findall(!=(0), Z[:, c_s] - Z[:, c_p])
```

Comment:
The column difference could be stored in a pre-calculated vector from outside the loop like
```julia
@. tmp = @views Z[:, c_2] - Z[:, c_p]
```

Reply:
More generally, seems like this could be written as a non-allocating loop that bails if a second match is found instead of continuing the search, and if a second match isn't found saves it in `robust_species`.

Reply:
Re the first comment: would I accomplish the same thing by adding `@views` in the `supp` line, so it becomes
```julia
supp = findall(!=(0), Z[:, c_s] - Z[:, c_p]
```
Seems like allocating it entirely outside might take a lot of space (would have to be a 3D array, where each difference vector is indexed by (c_s, c_p))

For the second comment, what do you mean by second match?